Refactor and improve AOTAutograd tests

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #103198
* __->__ #103197
* #103196

This is in preparation for the new "custom_op_compile_check" utility,
which will call the refactored testing API as a subroutine.

Here are the improvements to the AOTAutograd tests that this PR makes:
- we use torch.autograd.grad instead of .backward(), which makes it so
that we stop destructively modifying the inputs
- we get rid of the difficult-to-understand sentinel=42 logic and
replace it with something more sane
- We create some helper functions and add some code comments
- We improve error messages

Test Plan:
- wait for CI

zou3519 committed:
Refactor and improve AOTAutograd tests

This is in preparation for the new "custom_op_compile_check" utility,
which will call the refactored testing API as a subroutine.

Here are the improvements to the AOTAutograd tests that this PR makes:
- we use torch.autograd.grad instead of .backward(), which makes it so
that we stop destructively modifying the inputs
- we get rid of the difficult-to-understand sentinel=42 logic and
replace it with something more sane
- We create some helper functions and add some code comments
- We improve error messages

Test Plan:
- wait for CI

[ghstack-poisoned]

zou3519 committed:
Update on "Refactor and improve AOTAutograd tests"


This is in preparation for the new "custom_op_compile_check" utility,
which will call the refactored testing API as a subroutine.

Here are the improvements to the AOTAutograd tests that this PR makes:
- we use torch.autograd.grad instead of .backward(), which makes it so
that we stop destructively modifying the inputs
- we get rid of the difficult-to-understand sentinel=42 logic and
replace it with something more sane
- We create some helper functions and add some code comments
- We improve error messages

Test Plan:
- wait for CI

[ghstack-poisoned]

zou3519 committed:
Update on "Refactor and improve AOTAutograd tests"


This is in preparation for the new "custom_op_compile_check" utility,
which will call the refactored testing API as a subroutine.

Here are the improvements to the AOTAutograd tests that this PR makes:
- we use torch.autograd.grad instead of .backward(), which makes it so
that we stop destructively modifying the inputs
- we get rid of the difficult-to-understand sentinel=42 logic and
replace it with something more sane
- We create some helper functions and add some code comments
- We improve error messages

Test Plan:
- wait for CI

[ghstack-poisoned]

zou3519 committed:
Update on "Refactor and improve AOTAutograd tests"


This is in preparation for the new "custom_op_compile_check" utility,
which will call the refactored testing API as a subroutine.

Here are the improvements to the AOTAutograd tests that this PR makes:
- we use torch.autograd.grad instead of .backward(), which makes it so
that we stop destructively modifying the inputs
- we get rid of the difficult-to-understand sentinel=42 logic and
replace it with something more sane
- We create some helper functions and add some code comments
- We improve error messages

Test Plan:
- wait for CI

[ghstack-poisoned]

bdhirsh commented on torch/testing/_internal/optests/aot_autograd.py:
```diff
@@ -0,0 +1,115 @@
+    def call_forwards_backwards(f, args):
+        flat_args, _ = pytree.tree_flatten(args)
+        diff_args = [arg for arg in flat_args if isinstance(arg, torch.Tensor) and
+                     arg.requires_grad and arg.is_leaf]
```
Probably not a big deal, but is there any reason to have the `arg.is_leaf` check in here? Now that we're using `autograd.grad()`, we can just get gradients w.r.t. all inputs that require grad (leaf or not) and make sure they're right.

The example in my head is that we have a bunch of AOTAutograd tests that test input mutations. You can't mutate leaf tensors, so those tests create a leaf, do some compute on it, and pass that (non-leaf) computed tensor as an input to aot autograd. Those tests don't use this util though, so this probably doesn't really matter.

zou3519 replied:
Yes, you're right that we don't need the arg.is_leaf check. I wrote this code before I modified the logic to use autograd.grad instead of .backward() and didn't update it; this is another thing that is nice about switching to torch.autograd.grad for testing

soulitzer commented on torch/testing/_internal/optests/aot_autograd.py:
```diff
@@ -0,0 +1,115 @@
+                sm += i.sum().abs()
+            sm
+        else:
+            sm = out.sum().abs()
```
Looks like this was here before, but any idea why this abs() is here?

zou3519 replied:
Good question. I dug into this a bit, and it looks like the result of out.sum() could be a complex scalar output (https://github.com/pytorch/pytorch/pull/92753). I'll add a comment about it.

soulitzer commented on torch/testing/_internal/optests/aot_autograd.py:
```diff
@@ -0,0 +1,115 @@
+            sm = 0
+            for i in flat_out:
+                sm += i.sum().abs()
+            sm
```
typo?

zou3519 committed:
Update on "Refactor and improve AOTAutograd tests"


This is in preparation for the new "custom_op_compile_check" utility,
which will call the refactored testing API as a subroutine.

Here are the improvements to the AOTAutograd tests that this PR makes:
- we use torch.autograd.grad instead of .backward(), which makes it so
that we stop destructively modifying the inputs
- we get rid of the difficult-to-understand sentinel=42 logic and
replace it with something more sane
- We create some helper functions and add some code comments
- We improve error messages

Test Plan:
- wait for CI

[ghstack-poisoned]
